pkg/acpi/acpi.go

```diff
@@ -0,0 +1,219 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package acpi began life as a relatively simple set of functions.
+// At some point we wanted to be able to extend the ACPI table over a
+// kexec and that's when it all went nasty.
+// A few issues crop up.
+// In theory, one should be able to put tables anywhere. In practice,
+// even in the EFI age, this is not simple. The root pointer
+// for tables, RSDP, Root Services Data Pointer, is in memory protected by
+// the southbridge/ICH/PCH with lockdown bits (disable writes once, can
+// not be reset until power on/reset). Writes to it are transparently discarded.
+// (This makes sense, as crap DOS programs probably tried to write that area all
+// the time. Everything -- everything! -- about x86 and BIOS is explainable
+// by DOS, 16 bit addresses, 4 bit segments, 20 bits, and the incredible failure
+// of vision in 1999 of building a bunch of tables with 32 bit pointers in them.
+// The x86 world is nothing if not consistent: short-sighted decisions for 40+ years).
+// The two-byte EBDA pointer (16 bits) is shifted left 4 bits to get a 20-bit address
+// which points to *the range of memory* containing, maybe, the RSDP.
+// The RSDP has to live in the low 20 bits of address space (remember 20 bits, right?)
+// (Yes, I know about the EFI tables which make this restriction less of a problem,
+//  but for now all the firmware I've seen adheres to the "RSDP in e or f segment" rule.)
+// The RSDP has two pointers in it, one or both containing a value: 32 bit pointer or 64 bit
+// pointer to the RSDT or XSDT. (see "failure of vision" above.)
+// RSDT has 32-bit pointers, XSDT 64-bit pointers.
+// The spec recommends you ignore the 32-bit pointer variants but as of 2019,
+// I still see use of them; this package will read and write the variants but the
+// internal structs are defined with 64-bit pointers (oh, if only the ACPICA code
+// had gotten this right, but that code is pretty bad too).
+// What's really fun: when you generate the [RX]SDT, you need to generate pointers
+// too. It's best to keep those tables in a physically contiguous area, so we do;
+// not all systems follow this rule, which is asking for trouble.
+// Finally, not all tables are generated in a consistent way, e.g. the IBFT is not like
+// most other tables containing variable length data, because the people who created it
+// are Klever. IBFT has its own heap, and the elements in the heap follow rules unlike
+// all other tables. Nice!
+// If you really look at ACPI closely, you can kind of see that it's optimized for NASM,
+// which is part of what makes it so unpleasant. But, it's what we have.
+package acpi
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"log"
+	"strconv"
+)
+
+const (
+	// LengthOffset is the offset of the table length
+	LengthOffset = 4
+	// CSUMOffset is the offset of the single byte checksum in *most* ACPI tables
```

Comment:
This scares me a bit when you say "most". Would it make more sense for CSUMOffset to be a function in the Tabler interface?

Reply:
I don't think so.